File: DPipeSourceCPP/impl/dpmessangerbase.cpp
#include "dpmessangerbase.h"

using namespace std;
using namespace crdk::dpipes;

DPMessageReceivedAsyncData::DPMessageReceivedAsyncData(
	IDPipeMessanger* messangerPtr,
	PacketHeader packetHeader,
	std::shared_ptr<HeapAllocatedData> heapData,
	IDPipe* dpipe) :
	header(false, packetHeader.DataSize()) {
	header.SetDataCode(packetHeader.GetDataCode());
	messanger = messangerPtr;
	data = heapData;
	pipe = dpipe;
}

SendDataAsyncContainer::SendDataAsyncContainer(
	IDPipeMessanger* messangerPtr, 
	DP_MESSAGE_TYPE messageType, 
	void* dataPtr, 
	DWORD dataSize, 
	std::function<void(IDPipe*)> callbackFunction) {
	dpMessanger = messangerPtr;
	type = messageType;
	data = dataPtr;
	size = dataSize;
	callback = callbackFunction;
}

SendStringAsyncContainer::SendStringAsyncContainer(
	IDPipeMessanger* messangerPtr,
	DP_MESSAGE_TYPE messageType,
	string dataStr,
	function<void(IDPipe*)> callbackFunction) {
	dpMessanger = messangerPtr;
	type = messageType;
	data = dataStr;
	callback = callbackFunction;
}

SendWStringAsyncContainer::SendWStringAsyncContainer(
	IDPipeMessanger* messangerPtr,
	DP_MESSAGE_TYPE messageType,
	wstring dataStr,
	function<void(IDPipe*)> callbackFunction) {
	dpMessanger = messangerPtr;
	type = messageType;
	data = dataStr;
	callback = callbackFunction;
}


IDPipeMessanger::IDPipeMessanger(IDPipe* dpipe) {
	_dpipe = dpipe;
}

IDPipeMessanger::~IDPipeMessanger() {

}

void IDPipeMessanger::OnMessageReceived(IDPipe* pipe, PacketHeader& header, std::shared_ptr<HeapAllocatedData> heapData) {

	switch (header.GetDataCodeOnly())
	{
	case DP_MESSAGE_DATA:
		if (_onMessageDataReceived)
			_onMessageDataReceived(pipe, header, heapData);
		break;
	case DP_INFO_DATA:
		if (_onInfoDataReceived)
			_onInfoDataReceived(pipe, header, heapData);
		break;
	case DP_WARNING_DATA:
		if (_onWarningDataReceived)
			_onWarningDataReceived(pipe, header, heapData);
		break;
	case DP_ERROR_DATA:
		if (_onErrorDataReceived)
			_onErrorDataReceived(pipe, header, heapData);
		break;
	}
}

void IDPipeMessanger::OnMessageReceivedAsync(LPVOID dataPtr) {
	auto messageData = (DPMessageReceivedAsyncData*)dataPtr;
	IDPipeMessanger* messanger = messageData->messanger;
	messanger->OnMessageReceived(messageData->pipe, messageData->header, messageData->data);
	delete messageData;
}

void IDPipeMessanger::OnMessageStringReceived(IDPipe* pipe, PacketHeader& header, std::shared_ptr<HeapAllocatedData> heapData) {
	
	//Get data code ignoring first 8 bits (using for encoding)
	switch (header.GetDataCodeOnly())
	{
	case DP_MESSAGE_STRING:
		if (_onMessageStringReceived)
			_onMessageStringReceived(pipe, header, heapData);
		break;
	case DP_INFO_STRING:
		if (_onInfoStringReceived)
			_onInfoStringReceived(pipe, header, heapData);
		break;
	case DP_WARNING_STRING:
		if (_onWarningStringReceived)
			_onWarningStringReceived(pipe, header, heapData);
		break;
	case DP_ERROR_STRING:
		if (_onErrorStringReceived)
			_onErrorStringReceived(pipe, header, heapData);
		break;
	}
}

void IDPipeMessanger::OnMessageStringReceivedAsync(LPVOID dataPtr) {
	auto messageData = (DPMessageReceivedAsyncData*)dataPtr;
	IDPipeMessanger* messanger = messageData->messanger;
	messanger->OnMessageStringReceived(messageData->pipe, messageData->header, messageData->data);
	delete messageData;
}

DWORD IDPipeMessanger::GetEncodingCode(PacketHeader header) {
	if (header.IsService())
		return header.GetServicePrefix();

	return header.GetDataPrefix();
}

DWORD IDPipeMessanger::GetEncodingCode(DP_STRING_TYPE stringType) {
	if (stringType == DP_STRING_TYPE::STRING)
		return DP_ENCODING_UTF8;
	else if (stringType == DP_STRING_TYPE::WSTRING)
		return DP_ENCODING_UNICODE;
	return DP_ENCODING_UTF8;
}

DWORD IDPipeMessanger::AddEncodingCode(DWORD serviceCode, DWORD encodingCode) {
	auto encodingRaw = encodingCode << 24;
	return serviceCode | encodingRaw;
}

DP_STRING_TYPE IDPipeMessanger::GetStringType(DWORD encodingCode)
{
	if (encodingCode == DP_ENCODING_UTF8)
		return DP_STRING_TYPE::STRING;
	else if (encodingCode == DP_ENCODING_UNICODE)
		return DP_STRING_TYPE::WSTRING;
	else
		return DP_STRING_TYPE::UNDEFINED;
}

DP_STRING_TYPE IDPipeMessanger::GetStringType(PacketHeader header) {
	return GetStringType(GetEncodingCode(header));
}

wstring IDPipeMessanger::StringToWString(const std::string& str) {
	wstring wstr;
	size_t size;
	wstr.resize(str.length());
	mbstowcs_s(&size, &wstr[0], wstr.size() + 1, str.c_str(), str.size());
	return wstr;
}

string IDPipeMessanger::WStringToString(const std::wstring& wstr)
{
	string str;
	size_t size;
	str.resize(wstr.length());
	wcstombs_s(&size, &str[0], str.size() + 1, wstr.c_str(), wstr.size());
	return str;
}


string IDPipeMessanger::GetString(PacketHeader header, std::shared_ptr<HeapAllocatedData> heapData) {

	if (header.DataSize() == 0)
		return string();

	if (GetStringType(header) == DP_STRING_TYPE::WSTRING) {
		wstring result((wchar_t*)heapData->data(), heapData->size() / sizeof(wchar_t));
		return WStringToString(result);
	}

	return string((char*)heapData->data(), heapData->size());
}

string IDPipeMessanger::GetString(PacketHeader header) {

	auto dataSize = header.DataSize();

	if (dataSize == 0)
		return string();

	HeapAllocatedData data(dataSize, false);
	_dpipe->Read(data.data(), dataSize);

	if (GetStringType(header) == DP_STRING_TYPE::WSTRING) {
		wstring result((wchar_t*)data.data(), dataSize / sizeof(wchar_t));
		return WStringToString(result);
	}

	return string((char*)data.data(), dataSize);
}

string IDPipeMessanger::GetString(std::shared_ptr<HeapAllocatedData> heapData, int encodingCode) {
	if (heapData->size() == 0)
		return string();

	if (GetStringType(encodingCode) == DP_STRING_TYPE::WSTRING) {
		wstring result((wchar_t*)heapData->data(), heapData->size() / sizeof(wchar_t));
		return WStringToString(result);
	}

	return string((char*)heapData->data(), heapData->size());
}

wstring IDPipeMessanger::GetWString(PacketHeader header, std::shared_ptr<HeapAllocatedData> heapData) {
	if (header.DataSize() == 0)
		return wstring();

	if (GetStringType(header) == DP_STRING_TYPE::STRING) {
		string result ((char*)heapData->data(), heapData->size());
		return StringToWString(result);
	}

	return wstring((wchar_t*)heapData->data(), heapData->size() / sizeof(wchar_t));
}

wstring IDPipeMessanger::GetWString(PacketHeader header) {

	auto dataSize = header.DataSize();

	if (dataSize == 0)
		return wstring();

	HeapAllocatedData data(dataSize, false);
	_dpipe->Read(data.data(), dataSize);

	if (GetStringType(header) == DP_STRING_TYPE::STRING) {
		string result((char*)data.data(), dataSize);
		return StringToWString(result);
	}

	return wstring((wchar_t*)data.data(), dataSize / sizeof(wchar_t));
}

wstring IDPipeMessanger::GetWString(std::shared_ptr<HeapAllocatedData> heapData, int encodingCode)
{
	if (heapData->size() == 0)
		return wstring();

	if (GetStringType(encodingCode) == DP_STRING_TYPE::STRING) {
		string result((char*)heapData->data(), heapData->size());
		return StringToWString(result);
	}

	return wstring((wchar_t*)heapData->data(), heapData->size() / sizeof(wchar_t));
}

bool IDPipeMessanger::Send(DP_MESSAGE_TYPE messageType, void* data, DWORD dataSize) {

	DWORD dataCode = 0;

	switch (messageType) {
		case DP_MESSAGE_TYPE::MESSAGE:
			dataCode = DP_MESSAGE_DATA;
			break;
		case DP_MESSAGE_TYPE::MESSAGE_INFO:
			dataCode = DP_INFO_DATA;
			break;
		case DP_MESSAGE_TYPE::MESSAGE_WARNING:
			dataCode = DP_WARNING_DATA;
			break;
		case DP_MESSAGE_TYPE::MESSAGE_ERROR:
			dataCode = DP_ERROR_DATA;
			break;
		default:
			throw exception("Unkonws message type");
	}

	DWORD nBytesWritten;

	_mutexWrite.lock();
		bool result = _dpipe->Write(dataCode, data, dataSize, &nBytesWritten);
	_mutexWrite.unlock();

	return result;
}

bool IDPipeMessanger::SendDataStatic(LPVOID dataPtr) {
	SendDataAsyncContainer* sendData = (SendDataAsyncContainer*)dataPtr;
	bool result = sendData->dpMessanger->Send(sendData->type, sendData->data, sendData->size);

	if (sendData->callback) {
		sendData->callback(sendData->dpMessanger->_dpipe);
	}

	return result;
}

bool IDPipeMessanger::SendAsync(DP_MESSAGE_TYPE messageType, void* data, DWORD dataSize, std::function<void(IDPipe* pipe)> callback) {
	SendDataAsyncContainer* sendData = new SendDataAsyncContainer(this, messageType, data, dataSize, callback);
	DWORD _nReadThreadId;
	auto _tHandleThread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)SendDataStatic, (LPVOID)sendData, 0, &_nReadThreadId);
	return _tHandleThread != INVALID_HANDLE_VALUE;
}

bool crdk::dpipes::IDPipeMessanger::Send(std::string message) {
	return Send(DP_MESSAGE_TYPE::MESSAGE, message);
}

bool IDPipeMessanger::Send(DP_MESSAGE_TYPE messageType, std::string message)
{
	DWORD dataCode = 0;

	switch (messageType) {
	case DP_MESSAGE_TYPE::MESSAGE:
		dataCode = AddEncodingCode(DP_MESSAGE_STRING, _stringEncodingCode);
		break;
	case DP_MESSAGE_TYPE::MESSAGE_INFO:
		dataCode = AddEncodingCode(DP_INFO_STRING, _stringEncodingCode);
		break;
	case DP_MESSAGE_TYPE::MESSAGE_WARNING:
		dataCode = AddEncodingCode(DP_WARNING_STRING, _stringEncodingCode);
		break;
	case DP_MESSAGE_TYPE::MESSAGE_ERROR:
		dataCode = AddEncodingCode(DP_ERROR_STRING, _stringEncodingCode);
		break;
	default:
		throw exception("Unkonws message type");
	}

	DWORD nBytesWritten;
	DWORD len = boost::numeric_cast<DWORD>(message.length());

	_mutexWrite.lock();
		bool result = _dpipe->Write(dataCode, message.data(), len, &nBytesWritten);
	_mutexWrite.unlock();

	return result;
}

bool IDPipeMessanger::SendStringStatic(LPVOID dataPtr) {
	SendStringAsyncContainer* sendData = (SendStringAsyncContainer*)dataPtr;
	bool result = sendData->dpMessanger->Send(sendData->type, sendData->data);

	if (sendData->callback) {
		sendData->callback(sendData->dpMessanger->_dpipe);
	}

	return result;
}

bool IDPipeMessanger::SendAsync(DP_MESSAGE_TYPE messageType, std::string message, std::function<void(IDPipe* pipe)> callback) {
	SendStringAsyncContainer* sendData = new SendStringAsyncContainer(this, messageType, message, callback);
	DWORD _nReadThreadId;
	auto _tHandleThread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)SendStringStatic, (LPVOID)sendData, 0, &_nReadThreadId);
	return _tHandleThread != INVALID_HANDLE_VALUE;
}

IDPipe* crdk::dpipes::IDPipeMessanger::GetPipe() const{
	return _dpipe;
}

bool crdk::dpipes::IDPipeMessanger::Send(std::wstring message) {
	return Send(DP_MESSAGE_TYPE::MESSAGE, message);
}

bool IDPipeMessanger::Send(DP_MESSAGE_TYPE messageType, std::wstring message)
{
	DWORD dataCode = 0;

	switch (messageType) {
	case DP_MESSAGE_TYPE::MESSAGE:
		dataCode = AddEncodingCode(DP_MESSAGE_STRING, _wstringEncodingCode);
		break;
	case DP_MESSAGE_TYPE::MESSAGE_INFO:
		dataCode = AddEncodingCode(DP_INFO_STRING, _wstringEncodingCode);
		break;
	case DP_MESSAGE_TYPE::MESSAGE_WARNING:
		dataCode = AddEncodingCode(DP_WARNING_STRING, _wstringEncodingCode);
		break;
	case DP_MESSAGE_TYPE::MESSAGE_ERROR:
		dataCode = AddEncodingCode(DP_ERROR_STRING, _wstringEncodingCode);
		break;
	default:
		throw exception("Unkonws message type");
	}

	DWORD nBytesWritten;
	DWORD len = boost::numeric_cast<DWORD>(message.length() * sizeof(wchar_t));

	_mutexWrite.lock();
		bool result = _dpipe->Write(dataCode, message.data(), len, &nBytesWritten);
	_mutexWrite.unlock();

	return result;

}

bool IDPipeMessanger::SendWStringStatic(LPVOID dataPtr) {
	SendWStringAsyncContainer* sendData = (SendWStringAsyncContainer*)dataPtr;
	bool result = sendData->dpMessanger->Send(sendData->type, sendData->data);

	if (sendData->callback) {
		sendData->callback(sendData->dpMessanger->_dpipe);
	}

	return result;
}

bool IDPipeMessanger::SendAsync(DP_MESSAGE_TYPE messageType, std::wstring message, std::function<void(IDPipe* pipe)> callback) {
	SendWStringAsyncContainer* sendData = new SendWStringAsyncContainer(this, messageType, message, callback);
	DWORD _nReadThreadId;
	auto _tHandleThread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)SendWStringStatic, (LPVOID)sendData, 0, &_nReadThreadId);
	return _tHandleThread != INVALID_HANDLE_VALUE;
}

bool IDPipeMessanger::SendMessageString(const std::string message) {
	return Send(DP_MESSAGE_TYPE::MESSAGE, message);
}

bool IDPipeMessanger::SendMessageString(const std::wstring message) {
	return Send(DP_MESSAGE_TYPE::MESSAGE, message);
}

bool IDPipeMessanger::SendMessageStringAsync(const std::string message, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE, message, callback);
}

bool IDPipeMessanger::SendMessageStringAsync(const std::wstring message, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE, message, callback);
}

bool IDPipeMessanger::SendMessageData(void* data, DWORD dataSize) {
	return Send(DP_MESSAGE_TYPE::MESSAGE, data, dataSize);
}

bool IDPipeMessanger::SendMessageDataAsync(void* data, DWORD dataSize, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE, data, dataSize, callback);
}

bool IDPipeMessanger::SendInfo(const std::string message){
	return Send(DP_MESSAGE_TYPE::MESSAGE_INFO, message);
}

bool IDPipeMessanger::SendInfo(const std::wstring message) {
	return Send(DP_MESSAGE_TYPE::MESSAGE_INFO, message);
}

bool IDPipeMessanger::SendInfoAsync(const std::string message, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE_INFO, message);
}

bool IDPipeMessanger::SendInfoAsync(const std::wstring message, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE_INFO, message);
}

bool IDPipeMessanger::SendInfo(void* data, DWORD dataSize) {
	return Send(DP_MESSAGE_TYPE::MESSAGE_INFO, data, dataSize);
}

bool IDPipeMessanger::SendInfoAsync(void* data, DWORD dataSize, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE_INFO, data, dataSize, callback);;
}

bool IDPipeMessanger::SendWarning(const std::string message) {
	return Send(DP_MESSAGE_TYPE::MESSAGE_WARNING, message);
}

bool IDPipeMessanger::SendWarning(const std::wstring message) {
	return Send(DP_MESSAGE_TYPE::MESSAGE_WARNING, message);
}

bool IDPipeMessanger::SendWarningAsync(const std::string message, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE_WARNING, message, callback);
}

bool IDPipeMessanger::SendWarningAsync(const std::wstring message, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE_WARNING, message, callback);
}

bool IDPipeMessanger::SendWarning(void* data, DWORD dataSize) {
	return Send(DP_MESSAGE_TYPE::MESSAGE_WARNING, data, dataSize);
}

bool IDPipeMessanger::SendWarningAsync(void* data, DWORD dataSize, std::function<void(IDPipe* pipe)> callback){
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE_WARNING, data, dataSize, callback);
}

bool IDPipeMessanger::SendError(const std::string message) {
	return Send(DP_MESSAGE_TYPE::MESSAGE_ERROR, message);
}

bool IDPipeMessanger::SendError(const std::wstring message) {
	return Send(DP_MESSAGE_TYPE::MESSAGE_ERROR, message);
}

bool IDPipeMessanger::SendErrorAsync(const std::string message, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE_ERROR, message, callback);
}

bool IDPipeMessanger::SendErrorAsync(const std::wstring message, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE_ERROR, message, callback);
}

bool IDPipeMessanger::SendError(void* data, DWORD dataSize) {
	return Send(DP_MESSAGE_TYPE::MESSAGE_ERROR, data, dataSize);
}

bool IDPipeMessanger::SendErrorAsync(void* data, DWORD dataSize, std::function<void(IDPipe* pipe)> callback) {
	return SendAsync(DP_MESSAGE_TYPE::MESSAGE_ERROR, data, dataSize, callback);
}

void IDPipeMessanger::SetMessageStringReceivedHandler(std::function<void(IDPipe* pipe, PacketHeader header, std::shared_ptr<HeapAllocatedData>data)> function) {
	_onMessageStringReceived = function;
}

void IDPipeMessanger::SetMessageDataReceivedHandler(std::function<void(IDPipe* pipe, PacketHeader header, std::shared_ptr<HeapAllocatedData>data)> function) {
	_onMessageDataReceived = function;
}

void IDPipeMessanger::SetInfoStringReceivedHandler(std::function<void(IDPipe* pipe, PacketHeader header, std::shared_ptr<HeapAllocatedData>data)> function) {
	_onInfoStringReceived = function;
}

void IDPipeMessanger::SetInfoDataReceivedHandler(std::function<void(IDPipe* pipe, PacketHeader header, std::shared_ptr<HeapAllocatedData>data)> function) {
	_onInfoDataReceived = function;
}

void IDPipeMessanger::SetWarningStringReceivedHandler(std::function<void(IDPipe* pipe, PacketHeader header, std::shared_ptr<HeapAllocatedData>data)> function) {
	_onWarningStringReceived = function;
}

void IDPipeMessanger::SetWarningDataReceivedHandler(std::function<void(IDPipe* pipe, PacketHeader header, std::shared_ptr<HeapAllocatedData>data)> function) {
	_onWarningDataReceived = function;
}

void IDPipeMessanger::SetErrorStringReceivedHandler(std::function<void(IDPipe* pipe, PacketHeader header, std::shared_ptr<HeapAllocatedData>data)> function) {
	_onErrorStringReceived = function;
}

void IDPipeMessanger::SetErrorDataReceivedHandler(std::function<void(IDPipe* pipe, PacketHeader header, std::shared_ptr<HeapAllocatedData>data)> function) {
	_onErrorDataReceived = function;
}

void IDPipeMessanger::SetStringEncodingCode(DWORD code) {
	_stringEncodingCode = code;
}

void IDPipeMessanger::SetWStringEncodingCode(DWORD code) {
	_wstringEncodingCode = code;
}

void IDPipeMessanger::Disconnect() {
	_dpipe->Disconnect();
}

void IDPipeMessanger::Disconnect(std::string message) {
	DWORD len = boost::numeric_cast<DWORD>(message.length());
	_dpipe->Disconnect(message.data(), len, DP_ENCODING_UTF8);
}

void IDPipeMessanger::Disconnect(std::wstring message) {
	DWORD len = boost::numeric_cast<DWORD>(message.length() * (sizeof(wchar_t)));
	_dpipe->Disconnect(message.data(), len, DP_ENCODING_UNICODE);
}

File: CreatorDK.IO.DPipes/DPMessangerBase.cs
using System;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace CreatorDK.IO.DPipes
{
    public enum DP_MESSAGE_TYPE
    {
        MESSAGE = 1,
        MESSAGE_INFO = 2,
        MESSAGE_WARNING = 3,
        MESSAGE_ERROR = 4
    }

    public class DPMessangerBase
    {
        protected readonly DPipe _dpipe;
        protected Encoding _encoding;
        protected Mutex _mutexWritePipe;

        public delegate void MessageStringRecevied(string message);
        public delegate void MessageDataRecevied(byte[] data);
        public Encoding Encoding 
        { 
            get 
            {
                if (_encoding == null)
                    _encoding = Encoding.Unicode;

                return _encoding; 
            }

            set
            {
                if (value == null)
                    _encoding = Encoding.Unicode;
                else
                    _encoding = value;
            }
        }
        public MessageStringRecevied OnMessageStringReceived { get; set; }
        public MessageDataRecevied OnMessageDataReceived { get; set; }
        public MessageStringRecevied OnInfoStringReceived { get; set; }
        public MessageDataRecevied OnInfoDataReceived { get; set; }
        public MessageStringRecevied OnWarningStringReceived { get; set; }
        public MessageDataRecevied OnWarningDataReceived { get; set; }
        public MessageStringRecevied OnErrorStringReceived { get; set; }
        public MessageDataRecevied OnErrorDataReceived { get; set; }
        public DPMessangerBase(DPipe dpipe, Encoding encoding, uint stringBufferSize = 4096)
        {
            _dpipe = dpipe;
            Encoding = encoding;
            _mutexWritePipe = new Mutex(false);
        }
        public static Encoding GetEncoding(uint encodingCode)
        {
            switch (encodingCode)
            {
                case 1: 
                    return Encoding.Unicode;
                default:
                    return Encoding.UTF8;
            }
        }
        public static uint GetEncodingCode(PacketHeader header)
        {
            if (header.IsService)
                return header.ServicePrefix;

            return header.DataPrefix;
        }
        public static uint GetEncodingCode(Encoding encoding)
        {
            if (encoding == null)
                return 1;

            if (encoding == Encoding.UTF8)
                return 0;

            return 1;
        }
        public Encoding GetEncoding(PacketHeader header)
        {
            var encodingCode = GetEncodingCode(header);
            return GetEncoding(encodingCode);
        }
        public static uint AddEncodingCode(uint serviceCode, uint encodingCode)
        {
            var encodingRaw = encodingCode << 24;
            return serviceCode | encodingRaw;
        }
        public virtual string GetString(PacketHeader header, Encoding encoding = null)
        {
            var buffer = new byte[header.DataSize];

            _dpipe.Read(buffer, 0, header.DataSize);

            if (encoding == null)
            {
                var encodingCode = GetEncodingCode(header);
                encoding = GetEncoding(encodingCode);
            }

            return encoding.GetString(buffer, 0, header.DataSize);
        }

        public string GetString(PacketHeader header, byte[] data)
        {
            return GetEncoding(header).GetString(data);
        }
        protected virtual void OnMessageDataReceivedInner(PacketHeader header, byte[] data)
        {
            //Get data code ignoring first 8 bits (using for encoding)
            switch (header.DataCodeOnly) 
            {
                case Constants.DP_MESSAGE_DATA:
                    OnMessageDataReceived?.Invoke(data);
                    break;
                case Constants.DP_INFO_DATA:
                    OnInfoDataReceived?.Invoke(data);
                    break;
                case Constants.DP_WARNING_DATA:
                    OnWarningDataReceived?.Invoke(data);
                    break;
                case Constants.DP_ERROR_DATA:
                    OnErrorDataReceived?.Invoke(data);
                    break;
            }
        }
        protected virtual void OnMessageStringReceivedInner(PacketHeader header, string message)
        {
            //Removing Encoding code from ServiceCode
            switch (header.DataCodeOnly)
            {
                case Constants.DP_MESSAGE_STRING:
                    OnMessageStringReceived?.Invoke(message);
                    break;
                case Constants.DP_INFO_STRING:
                    OnInfoStringReceived?.Invoke(message);
                    break;
                case Constants.DP_WARNING_STRING:
                    OnWarningStringReceived?.Invoke(message);
                    break;
                case Constants.DP_ERROR_STRING:
                    OnErrorStringReceived?.Invoke(message);
                    break;
            }
        }
        public virtual void SendMessage(byte[] data)
        {
            _mutexWritePipe.WaitOne();
                _dpipe.Write(Constants.DP_MESSAGE_DATA, data, 0, data.Length);
            _mutexWritePipe.ReleaseMutex();
        }
        public virtual void SendMessageAsync(byte[] data) 
        {
            Task.Run(() => SendMessage(data));
        }
        public virtual void SendMessage(string message, Encoding encoding = null)
        {
            var encodingCurrent = encoding == null ? Encoding : encoding;

            byte[] data = encodingCurrent.GetBytes(message);

            var code = AddEncodingCode(Constants.DP_MESSAGE_STRING, GetEncodingCode(encodingCurrent));

            _mutexWritePipe.WaitOne();
                _dpipe.Write(code, data, 0, data.Length);
            _mutexWritePipe.ReleaseMutex();
        }
        public virtual void SendMessageAsync(string message)
        {
            Task.Run(() => SendMessage(message));
        }
        public virtual void SendInfo(byte[] data)
        {
            _mutexWritePipe.WaitOne();
                _dpipe.Write(Constants.DP_INFO_DATA, data, 0, data.Length);
            _mutexWritePipe.ReleaseMutex();
        }
        public virtual void SendInfoAsync(byte[] data)
        {
            Task.Run(() => SendInfo(data));
        }
        public virtual void SendInfo(string message, Encoding encoding = null)
        {
            var encodingCurrent = encoding == null ? Encoding : encoding;

            byte[] data = encodingCurrent.GetBytes(message);

            var code = AddEncodingCode(Constants.DP_INFO_STRING, GetEncodingCode(encodingCurrent));

            _mutexWritePipe.WaitOne();
                _dpipe.Write(code, data, 0, data.Length);
            _mutexWritePipe.ReleaseMutex();
        }
        public virtual void SendInfoAsync(string message)
        {
            Task.Run(() => SendInfo(message));
        }
        public virtual void SendWarning(byte[] data)
        {
            _mutexWritePipe.WaitOne();
                _dpipe.Write(Constants.DP_WARNING_DATA, data, 0, data.Length);
            _mutexWritePipe.ReleaseMutex();
        }
        public virtual void SendWarningAsync(byte[] data)
        {
            Task.Run(() => SendWarning(data));
        }
        public virtual void SendWarning(string message, Encoding encoding = null)
        {
            var encodingCurrent = encoding == null ? Encoding : encoding;

            byte[] data = encodingCurrent.GetBytes(message);

            var code = AddEncodingCode(Constants.DP_WARNING_STRING, GetEncodingCode(encodingCurrent));

            _mutexWritePipe.WaitOne();
                _dpipe.Write(code, data, 0, data.Length);
            _mutexWritePipe.ReleaseMutex();
        }
        public virtual void SendWarningAsync(string message)
        {
            Task.Run(() => SendWarning(message));
        }
        public virtual void SendError(byte[] data)
        {
            _mutexWritePipe.WaitOne();
                _dpipe.Write(Constants.DP_ERROR_DATA, data, 0, data.Length);
            _mutexWritePipe.ReleaseMutex();
        }
        public virtual void SendErrorAsync(byte[] data)
        {
            Task.Run(() => SendError(data));
        }
        public virtual void SendError(string message, Encoding encoding = null)
        {
            var encodingCurrent = encoding == null ? Encoding : encoding;

            byte[] data = encodingCurrent.GetBytes(message);

            var code = AddEncodingCode(Constants.DP_WARNING_STRING, GetEncodingCode(encodingCurrent));

            _mutexWritePipe.WaitOne();
                _dpipe.Write(code, data, 0, data.Length);
            _mutexWritePipe.ReleaseMutex();
        }
        public virtual void SendErrorAsync(string message)
        {
            Task.Run(() => SendError(message));
        }

        public void Send(DP_MESSAGE_TYPE messageType, byte[] data)
        {
            switch(messageType) 
            { 
                case DP_MESSAGE_TYPE.MESSAGE:
                    SendMessage(data); break;
                case DP_MESSAGE_TYPE.MESSAGE_INFO:
                    SendInfo(data); break;
                case DP_MESSAGE_TYPE.MESSAGE_WARNING:
                    SendWarning(data); break;
                case DP_MESSAGE_TYPE.MESSAGE_ERROR:
                    SendError(data); break;
                default:
                    throw new Exception("Unknown message type");
            }
        }
        public void SendAsync(DP_MESSAGE_TYPE messageType, byte[] data)
        {
            Task.Run(() => Send(messageType, data));
        }

        public void Send(DP_MESSAGE_TYPE messageType, string message, Encoding encoding = null)
        {
            switch (messageType)
            {
                case DP_MESSAGE_TYPE.MESSAGE:
                    SendMessage(message, encoding); 
                    break;
                case DP_MESSAGE_TYPE.MESSAGE_INFO:
                    SendInfo(message, encoding); 
                    break;
                case DP_MESSAGE_TYPE.MESSAGE_WARNING:
                    SendWarning(message, encoding); 
                    break;
                case DP_MESSAGE_TYPE.MESSAGE_ERROR:
                    SendError(message, encoding); 
                    break;
                default:
                    throw new Exception("Unknown message type");
            }
        }
        public void SendAsync(DP_MESSAGE_TYPE messageType, string message, Encoding encoding = null)
        {
            Task.Run(() => Send(messageType, message, encoding));
        }

        public void Disconnect()
        {
            _dpipe.Disconnect();
        }
        public virtual void Disconnect(string disconnectMessage, Encoding encoding = null)
        {
            if (disconnectMessage == null)
                Disconnect();

            encoding = encoding == null ? Encoding : encoding;

            byte[] disonnectData = null;

            if (!string.IsNullOrEmpty(disconnectMessage))
                disonnectData = encoding.GetBytes(disconnectMessage);

            var encodingCode = GetEncodingCode(encoding);

            _dpipe.Disconnect(disonnectData, encodingCode);
        }
    }
}
